Remove CODEC DMIC for specific SKUs

charleskeepax opened the pull request:
Some SKUs are using host DMICs rather than the cs42l43 DMICs, update the endpoint structure for these SKUs to remove the DMIC endpoint. Longer term a more automated solution will be sensible, however for now this is limited to SKUs with a particular bus layout so this fix is easy to deploy whilst we work on a more complete solution.

plbossart commented on sound/soc/intel/common/soc-acpi-intel-lnl-match.c:
```diff
@@ -117,6 +117,27 @@ static const struct snd_soc_acpi_endpoint cs42l43_endpoints[] = {
+		.group_position = 0,
+		.group_id = 0,
+	},
+};
```
alternatively reshuffle the endpoint order and put DMICs last, then play with ARRAY_SIZE  - 1.

bardliao replied:
Or use a dai quick. Like https://github.com/thesofproject/linux/pull/4848/commits/47bae326da237ca6cf0414dc5a8a823d1ff83a39

charleskeepax replied:
Yeah I did kinda want to use a quirk, but it was a little awkward, since the polarity of the quirks is such that they are saying include this DAI. Which means we would need to add the quirk to every laptop except this one and given generally I expect the vast majority of SKUs to use the cs42l43 DMICs that feels suboptimal. I will have another look today see if I can work out a better way to invert the quirk polarity.

charleskeepax committed:
ASoC: Intel: sof_sdw: Add quirks for some new Lenovo laptops

Add some more sidecar amplifier quirks.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: sdw_utils: Add support for exclusion DAI quirks

The system contains a mechanism for certain DAI links to be included
based on a quirk. Add support for certain DAI links to excluded based on
a quirk, this is useful in situations where the vast majority of SKUs
utilise a feature so it is easier to quirk on those that don't.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

bardliao commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -1112,8 +1114,9 @@ int asoc_sdw_parse_sdw_endpoints(struct snd_soc_card *card,
-				if (dai_info->quirk && !(dai_info->quirk & ctx->mc_quirk))
-					continue;
+				if (dai_info->quirk &&
+				    !(dai_info->quirk_exclude ^ (dai_info->quirk & ctx->mc_quirk)))
```
Can we use `!dai_info->quirk_exclude * (dai_info->quirk & ctx->mc_quirk))`? It seems explain the `quirk_exclude` flag more.

charleskeepax replied:
I am not sure that expression results in the correct results, you want the expression to be true for any DAI that should be dropped for which there are two situations:

A DAI that has an include quirk, which is not specified in the machine quirks:
dai_info->quirk_exclude = 0, dai_info->quirk = 1, ctx->mc_quirk = 0
For this your expression would be: !0 * (1 & 0) = 0

A DAI that has an exclude quirk, that is specified in the machine quirks:
dai_info->quirk_exclude = 1, dai_info->quirk = 1, ctx->mc_quirk = 1
For this your expression would be: !1 * (1 & 1) = 0

I think the XOR is probably the best way to convey the intention. One could potentially be more explicit by expanding it out to something like:

if (dai_info->quirk &&
    ((!dai_info->quirk_exclude && !(dai_info->quirk & ctx->mc_quirk)) ||
    (dai_info->quirk_exclude && (dai_info->quirk & ctx->mc_quirk)))) {

But I think I prefer the shortened version.

bardliao replied:
Oh yes, thanks for the explanation. I thought we just need to revert the current result when dai_info->quirk_exclude = 1.

bardliao commented on include/sound/soc_sdw_utils.h:
```diff
@@ -28,6 +28,7 @@
  *   - SOC_SDW_CODEC_SPKR | SOF_SIDECAR_AMPS - Not currently supported
  */
 #define SOC_SDW_SIDECAR_AMPS		BIT(16)
+#define SOC_SDW_CODEC_NO_MIC		BIT(17)
```
Maybe just `SOC_SDW_CODEC_MIC` without `NO`? We have `quirk_exclude` to represent default value with or without the device. So we don't need to represent it in the quirk name, and save some quirk bits. I.e. Other codecs can use the `SOC_SDW_CODEC_MIC` quirk to don't use codec mic by default.

bardliao replied:
Also, can we split it to 2 commits? One to add the `quirk_exclude` flag and the other to make the cs42l43 MIC optional.

charleskeepax replied:
Yeah that makes sense will do that.

charleskeepax committed:
ASoC: sdw_utils: Add a quirk to allow the cs42l43 mic DAI to be ignored

To support some systems using host microphones add a quirk to allow the
cs42l43 microphone DAI link to be ignored.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: Intel: sof_sdw: Add quirk for cs42l43 system using host DMICs

Add quirk to inform the machine driver to not bind in the cs42l43
microphone DAI link.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax commented:
New version with the pieces split out, the define changes and the actual quirk added, now I have permission.

bardliao commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -599,6 +599,14 @@ static const struct dmi_system_id sof_sdw_quirk_table[] = {
+			DMI_MATCH(DMI_SYS_VENDOR, "LENOVO"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "233C")
+		},
+		.driver_data = (void *)(SOC_SDW_CODEC_MIC),
```
nitpick: I would recommend adding a comment to say the codec MIC is excluded. Because the quirk usually means included.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -487,6 +487,8 @@ struct asoc_sdw_codec_info codec_info_list[] = {
 				.widgets = generic_dmic_widgets,
 				.num_widgets = ARRAY_SIZE(generic_dmic_widgets),
+				.quirk = SOC_SDW_CODEC_MIC,
+				.quirk_exclude = true,
```
typo in commit message: micrphone

charleskeepax commented:
Sorry Bard timing was bad on that with respect to your approval there, was just fxing up the last checkpatch thing (one of those no commit message ones) and adding a new patch to add a couple of other quirks since we might as well get them all merged together.

charleskeepax commented:
Are we ok to get this one merged? The customer has been chasing us on it.
